index.js

```diff
@@ -24,221 +25,157 @@ function GotError(message, nested) {
-
-	// returns a proxy stream to the response
-	// if no callback has been provided
-	if (!cb) {
-		proxy = duplexify();
-
-		// forward errors on the stream
-		cb = function (err, data, response) {
-			proxy.emit('error', err, data, response);
-		};
-	}
-
-	if (proxy && json) {
-		throw new GotError('got can not be used as stream when options.json is used');
-	}
-
-	function get(opts, cb) {
+	function get(opts) {
 		var fn = opts.protocol === 'https:' ? https : http;
 		var url = urlLib.format(opts);
 
-		if (opts.agent === undefined) {
-			opts.agent = infinityAgent[fn === https ? 'https' : 'http'].globalAgent;
-
-			if (process.version.indexOf('v0.10') === 0 && fn === https && (
-				typeof opts.ca !== 'undefined' ||
-				typeof opts.cert !== 'undefined' ||
-				typeof opts.ciphers !== 'undefined' ||
-				typeof opts.key !== 'undefined' ||
-				typeof opts.passphrase !== 'undefined' ||
-				typeof opts.pfx !== 'undefined' ||
-				typeof opts.rejectUnauthorized !== 'undefined')) {
-				opts.agent = new infinityAgent.https.Agent({
-					ca: opts.ca,
-					cert: opts.cert,
-					ciphers: opts.ciphers,
-					key: opts.key,
-					passphrase: opts.passphrase,
-					pfx: opts.pfx,
-					rejectUnauthorized: opts.rejectUnauthorized
-				});
-			}
-		}
-
-		var req = fn.request(opts, function (response) {
-			var statusCode = response.statusCode;
-			var res = response;
-
-			// auto-redirect only for GET and HEAD methods
+		var req = fn.request(opts, function (res) {
+			var statusCode = res.statusCode;
 			if (isRedirect(statusCode) && 'location' in res.headers && (opts.method === 'GET' || opts.method === 'HEAD')) {
-				// discard response
 				res.resume();
 
 				if (++redirectCount > 10) {
-					cb(new GotError('Redirected 10 times. Aborting.'), undefined, res);
+					ee.emit('error', new GotError('Redirected 10 times. Aborting.'), undefined, res);
 					return;
 				}
 
 				var redirectUrl = urlLib.resolve(url, res.headers.location);
 				var redirectOpts = objectAssign({}, opts, urlLib.parse(redirectUrl));
 
-				if (opts.agent === infinityAgent.http.globalAgent && redirectOpts.protocol === 'https:' && opts.protocol === 'http:') {
-					redirectOpts.agent = undefined;
-				}
-
-				if (proxy) {
-					proxy.emit('redirect', res, redirectOpts);
-				}
+				ee.emit('redirect', res, redirectOpts);
 
-				get(redirectOpts, cb);
+				get(redirectOpts);
 				return;
 			}
 
-			if (proxy) {
-				proxy.emit('response', res);
-			}
-
 			if (['gzip', 'deflate'].indexOf(res.headers['content-encoding']) !== -1) {
-				res = res.pipe(zlib.createUnzip());
+				var unzip = zlib.createUnzip();
+				unzip.httpVersion = res.httpVersion;
+				unzip.headers = res.headers;
+				unzip.rawHeaders = res.rawHeaders;
+				unzip.trailers = res.trailers;
+				unzip.rawTrailers = res.rawTrailers;
+				unzip.setTimeout = res.setTimeout.bind(res);
+				unzip.statusCode = res.statusCode;
+				unzip.statusMessage = res.statusMessage;
+				unzip.socket = res.socket;
```

Comment:
It will copy internal fields - which breaks unzip stream. Maybe turn it to a module?